genie-common/src/main/java/com/netflix/genie/common/dto/v4/AgentEvent.java

```diff
@@ -0,0 +1,233 @@
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.common.dto.v4;
+
+import com.google.common.collect.Lists;
+import com.netflix.genie.common.dto.JobStatus;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * AgentEvent DTOs.
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Getter
+@EqualsAndHashCode(doNotUseGetters = true)
+@ToString(doNotUseGetters = true)
+public abstract class AgentEvent {
+    @NotEmpty
+    private final String agentId;
+    @NotNull
+    private final Instant timestamp;
+
+    private AgentEvent(
+        @NotBlank final String agentId,
+        final Instant... timestamp
+    ) {
+        this.agentId = agentId;
+        if (timestamp.length == 0) {
+            this.timestamp = Instant.now();
+        } else if (timestamp.length > 1) {
+            throw new IllegalArgumentException("Expecting 1 optional argument: timestamp, got " + timestamp.length);
+        } else {
+            this.timestamp = timestamp[0];
+        }
+    }
+
+    /**
+     * Event fired when a running job changes status.
+     */
+    @Getter
+    @EqualsAndHashCode(doNotUseGetters = true, callSuper = true)
+    @ToString(doNotUseGetters = true, callSuper = true)
+    public static class JobStatusUpdate extends AgentEvent {
+
+        @NotBlank
+        private final String jobId;
+        @NotNull
+        private final JobStatus jobStatus;
+
+        /**
+         * Constructor.
+         * @param agentId agent id
+         * @param jobId job id
+         * @param jobStatus current job status
+         * @param timestamp optional timestamp
+         */
+        public JobStatusUpdate(
+            final @NotBlank String agentId,
+            final @NotBlank String jobId,
+            final JobStatus jobStatus,
+            final Instant... timestamp
+        ) {
+            super(agentId, timestamp);
+            this.jobId = jobId;
+            this.jobStatus = jobStatus;
+        }
+    }
+
+    /**
+     * Event fired when the agent state machine changes state.
+     */
+    @Getter
+    @EqualsAndHashCode(doNotUseGetters = true, callSuper = true)
+    @ToString(doNotUseGetters = true, callSuper = true)
+    public static class StateChange extends AgentEvent {
+
+        @Nullable @NotEmpty
```

Comment:
What i meant to say here is: this can be null or it must be a non-empty string.
But i forgot to check wether the validator lets me get away with this. I'll check and probably change this.